src/core/writer.js

```diff
@@ -25,36 +25,84 @@ import { SimpleDOMNode, SimpleXMLParser } from "./xml_parser.js";
 import { BaseStream } from "./base_stream.js";
 import { calculateMD5 } from "./crypto.js";
 
-function writeObject(ref, obj, buffer, transform) {
+async function writeObject(ref, obj, buffer, transform) {
   buffer.push(`${ref.num} ${ref.gen} obj\n`);
   if (obj instanceof Dict) {
-    writeDict(obj, buffer, transform);
+    await writeDict(obj, buffer, transform);
   } else if (obj instanceof BaseStream) {
-    writeStream(obj, buffer, transform);
+    await writeStream(obj, buffer, transform);
   }
   buffer.push("\nendobj\n");
 }
 
-function writeDict(dict, buffer, transform) {
+async function writeDict(dict, buffer, transform) {
   buffer.push("<<");
   for (const key of dict.getKeys()) {
     buffer.push(` /${escapePDFName(key)} `);
-    writeValue(dict.getRaw(key), buffer, transform);
+    await writeValue(dict.getRaw(key), buffer, transform);
   }
   buffer.push(">>");
 }
 
-function writeStream(stream, buffer, transform) {
+async function writeStream(stream, buffer, transform) {
   let string = stream.getString();
   if (transform !== null) {
     string = transform.encryptString(string);
   }
+
+  // eslint-disable-next-line no-undef
+  if (typeof CompressionStream === "undefined") {
+    stream.dict.set("Length", string.length);
+    await writeDict(stream.dict, buffer, transform);
+    buffer.push(" stream\n", string, "\nendstream");
+    return;
+  }
+
+  const filter = stream.dict.getRaw("Filter");
+  const flateDecode = Name.get("FlateDecode");
+
+  // If the string is too small there is no real benefit
+  // in compressing it.
+  // The number 256 is arbitrary, but it should be reasonable.
+  const MIN_LENGTH_FOR_COMPRESSING = 256;
+
+  if (
+    string.length >= MIN_LENGTH_FOR_COMPRESSING ||
+    (Array.isArray(filter) && filter.includes(flateDecode)) ||
+    filter === flateDecode
```

Comment:
Given the `Name` caching, please change this into the following to prevent any future bugs.
```suggestion
    (filter instanceof Name && filter.name === flateDecode.name)
```

Reply:
I thought that the cache was useful to have only one instance of a given Name, e.g. `Name.get("FlateDecode")` is used for whatever `/FlatDecode` we've in the pdf and in such a case `filter===flatdecode` should be enough.
And if my understanding is wrong, what's the point to make a lookup in a hashmap when we could just use a `new Name...` ?

Reply:
The caching does make sense, and will work in *most* cases. However please keep in mind that (all) caches can be cleared, e.g. during idle cleanup, and in that case it's no longer guaranteed that e.g. `Name`-objects match.

Consider the following order of events:
 - The stream-dict /Filter entry is read and initialized somewhere else in the code-base, thus creating a new `Name`-instance and adding it to the `NameCache`. Furthermore, the dict-property is also being cached on the `XRef`-instance.
 - Some time passes, currently 30 seconds of inactivity in the viewer, which cause the `NameCache` to be cleared.
 - The user tries to save a modified document, and here the stream-dict /Filter entry now points to an *old* `Name`-instance (since the dict is cached on the `XRef`-instance). However, the `Name.get("FlateDecode");` call above creates a *new* `Name`-instance in the `NameCache` and we thus get a mismatch unless we compare the `name`-properties.

Obviously this situation will be really rare, but if/when it does happens we'll have a very difficult to find and highly intermittent bug on our hands.

Reply:
Ok, it makes sense.
What do you think of adding `Name.isEqual(n1, n2)` in Name with a comment (a copy/paste of the above comment will be enough I guess) explaining the reason for having it ?
It'll avoid to have to remember that (and to avoid the bus factor).

Reply:
Sure, that could work (although the comment should probably be generalized slightly compared).
Maybe just following the same pattern as in
https://github.com/mozilla/pdf.js/blob/4d3dfe254ffe90afff336ec28024e951b15ffad0/src/core/primitives.js#L372-L380
and 
https://github.com/mozilla/pdf.js/blob/4d3dfe254ffe90afff336ec28024e951b15ffad0/test/unit/primitives_spec.js#L579-L591

Reply:
Ok. Let's do that in a followup, would you mind to do it ?

Reply:
Sure, I'll try to look into that later.